### PR TITLE
use_self: Fix false positive in macro expansion

### LINKED ISSUE #6902: use_self still gives false positives
Now that #6818 is fixed and the fix (#6833) has made its way to the nightly there are still cases of false positives:

Lint name: `use_self`

I tried this code:

```rust
#![deny(clippy::use_self)]
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum Direction {
    North,
}

fn main() {
    let d = Direction::North;

    println!("{:?}", d);
}
```

I expected to see this happen: clippy clean - no warning/errors

Instead, this happened:

```
    Checking playground v0.0.1 (/playground)
error: unnecessary structure name repetition
 --> src/main.rs:6:6
  |
6 |   enum Direction {
  |  ______^
7 | |     North,
  | |____^ help: use the applicable keyword: `Self`
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(clippy::use_self)]
  |         ^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self

error: aborting due to previous error

error: could not compile `playground`

To learn more, run the command again with --verbose.
```

Interestingly enough it the code generated by `#[serde(untagged)]` that triggers this lint. If this line is commented out then clippy seems to be happy

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (2021-03-13 acca818)`

- `rustc -Vv`:
 ```
rustc 1.52.0-nightly (acca81892 2021-03-13)
binary: rustc
commit-hash: acca818928654807ed3bc1ce0e97df118f8716c8
commit-date: 2021-03-13
host: x86_64-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 12.0.0
 ```

### PR DESCRIPTION
**Pull request by Y-Nak**
fixes #6902 

changelog: `use_self`: Fix false positive related to macro expansion.

### CODE REVIEW

File: clippy_lints/src/use_self.rs [hunk 1]
```diff
@@ -467,3 +467,18 @@ fn should_lint_ty(hir_ty: &hir::Ty<'_>, ty: Ty<'_>, self_ty: Ty<'_>) -> bool {
...
+    }
+
+    for (parent_id, _) in map.parent_iter(hir_id) {
+        if map.opt_span(parent_id).map_or(false, in_macro) {
+            return true;
+        }
+    }
+
+    false
+}
```
Comment (camsteffen) [inline clippy_lints/src/use_self.rs:484]: This looks expensive and doesn't seem like the right solution to me. We probably need to check `in_macro` somewhere while descending into the HIR rather than the other way around.

Comment (flip1995) [reply clippy_lints/src/use_self.rs:484]: I agree, we shouldn't have to traverse the HIR to determine if something comes from a macro expansion. Is this maybe a bug in rustc, not tracking the origin of this span correctly?

**Comment by flip1995** [comment]
After doing some debugging I'm pretty sure this is a bug in the span lowering in rustc. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3c9dd6256ce48f86c16c1d345ff77596) (You can expand the code under "Tools").

Expanded code – relevant snippet:

```rust
if let _serde::__private::Ok(__ok) =
       match _serde::Deserializer::deserialize_any(_serde::__private::de::ContentRefDeserializer::<__D::Error>::new(&__content),
                                                   _serde::__private::de::UntaggedUnitVisitor::new("Direction",
                                                                                                   "North"))
           {
           _serde::__private::Ok(()) =>
           _serde::__private::Ok(Direction::North),
           _serde::__private::Err(__err) =>
           _serde::__private::Err(__err),
       } {
    return _serde::__private::Ok(__ok);
}
```

The error is emitted for `Direction::North`. The weird thing is that when removing `#[serde(untagged)]`  the exact same 

```rust
_serde::__private::Ok(Direction::North)
```

`Ctor` is used in the code, but doesn't trigger the lint.

`dbg!(span.ctxt().outer_expn_data()` for the `QPath` of `Direction::Nord` gives:

```
[clippy_lints/src/use_self.rs:398] span.ctxt().outer_expn_data() = ExpnData {
    kind: Root,
    parent: ExpnId(
        0,
    ),
    call_site: tests/ui/issue-6902.rs:1:1: 1:1 (#0),
    def_site: tests/ui/issue-6902.rs:1:1: 1:1 (#0),
    allow_internal_unstable: None,
    allow_internal_unsafe: false,
    local_inner_macros: false,
    edition: Edition2015,
    macro_def_id: Some(
        DefId(0:0 ~ issue_6902[317d]),
    ),
    krate: crate0,
    orig_id: Some(
        0,
    ),
    disambiguator: 0,
}
```

TL;DR: It doesn't come from an expansion at all. Since this code is obviously generated and clearly not written by the user the compiler should track that it is expanded. 

As a really wild guess, I would say, that this is because in the first case the expression is in the condition of an `if let` expression and in the second case it is just in a return position.

**Comment by flip1995** [comment]
So after trying to reproduce this for a few hours, I give up. I have this proc_macro that produces really similar code to the serde expansion:

```rust
#[proc_macro_derive(EnumUsedInIfLet)]
pub fn use_self(input: TokenStream) -> proc_macro::TokenStream {
    let item: syn::ItemEnum = parse_macro_input!(input);
    let _ident = item.ident;
    TokenStream::from(quote! {
        const _: () =
            {
                #[automatically_derived]
                impl Foo for #_ident {
                    fn foo() -> Result<Self, &'static str> {
                        if let Ok(ok) = match bar() {
                            Ok(()) => Ok(#_ident::A),
                            Err(err) => Err(err),
                        } {
                            return Ok(ok);
                        }
                        Err("Ohoh")
                    }
                }
            };
    })
}

```

Expanded:

```rust
#[derive(EnumUsedInIfLet)]
enum S { A }
```

```rust
enum S { A, }
const _: () =
    {
        #[automatically_derived]
        impl Foo for S {
            fn foo() -> Result<Self, &'static str> {
                if let Ok(ok) =
                       match bar() {
                           Ok(()) => Ok(S::A),
                           Err(err) => Err(err),
                       } {
                    return Ok(ok);
                }
                Err("Ohoh")
            }
        }
    };
```

But it doesn't trigger the lint. So my guess is, that serde does som string manipulation in the proc macro and therefore rustc is unable to tell that the expanded span comes from a macro.


---

So what I get from this: This fix doesn't work generally. It just randomly works here, because rustc is able to figure out the spans of the parent expressions.

**Comment by camsteffen** [comment]
Same root issue as #6514?

**Comment by flip1995** [comment]
> Unfortunately, neither Clippy nor rustc currently have a way to tell apart a quote_spanned span from a non-macro span, so this is currently impossible to fix on our side (Also see #6249 (comment)).

Seems like it

**Comment by Y-Nak** [comment]
Sorry for the late response, I took a vacation.
Yeah, I admit this is a really hacky solution and I'll close the PR.

In https://github.com/Y-Nak/clippy-6902-repro, I succeeded to find a minimal reproducer for #6902 and it seems not relevant to `quote_spanned`, so we may need further investigation on it.

> The weird thing is that when removing #[serde(untagged)] the exact same
_serde::__private::Ok(Direction::North)
Ctor is used in the code, but doesn't trigger the lint.

No, this is not weird because the `Direction::North` appears in `impl <'de> _serde::de::Visitor<'de> for __Visitor<'de>`  in the expansion.

**Comment by flip1995** [comment]
Ah thanks for doing some more investigation! Looking at your reproducer, I think the only thing I missed was that also the variant in the `qpath` had to be generated inside the macro with `syn`.

Do you want to open an issue in rust-lang/rust about this with your reproducer?

> No, this is not weird because the `Direction::North` appears in `impl <'de> _serde::de::Visitor<'de> for __Visitor<'de>` in the expansion.

Ah right, makes sense.

**Comment by Y-Nak** [comment]
> Do you want to open an issue in rust-lang/rust about this with your reproducer?

Yes, I'll open a PR.

**Comment by Y-Nak** [comment]
Opened https://github.com/rust-lang/rust/issues/84122.
